File: R/masking.R
source("R/path.R")
source("R/read.R")
library(docstring)
library(roxygen2)


apply_cond <- function(df, cond) {
  #' apply_cond
  #'
  #' Takes a pandas DataFrame and a condition, which can be a string, dictionary,
  #' or callable, and applies the condition to the DataFrame using \code{eval} or \code{apply}
  #' accordingly.
  #'
  #' @param df DataFrame. A pandas DataFrame containing the data on which the condition will be applied.
  #' @param cond MaskCondition. The condition to be applied on the dataframe. Can be either a string, a dictionary, or a
  #'             callable function.
  #'
  #' @return DataFrame. Dataframe evaluated at the mask condition.
  #'
  #' @export

  if (class(cond) == "character") {
    return(filter(eval(parse(text = cond))))
  } else if (class(cond) == "list") {
    cond <- paste(paste(names(cond), "==", paste0("'", unlist(cond), "'"), collapse = " & "), collapse = " & ")
    return(filter(eval(parse(text = cond))))
  } else if (inherits(cond, "function")) {
    return(apply(df, 1, cond))
  } else {
    stop("Unsupported condition type")
  }
}

#' Class for Masks
#'
#' @desctiption Class to define masks with conditions and weights to apply to DataFiles



Mask <- R6::R6Class("Mask",

    private = list(
        ..where = NULL,
        ..use = NULL,
        ..weight = NULL,
        ..other = NULL,
        ..comment = NULL

    ),
    public = list(
    #' @description
    #' Create a new mask object
    #' @param where MaskCondition | list[MaskCondition], optional. Where the mask should be applied.
    #' @param use MaskCondition | list[MaskCondition], optional. Condition on where to use the masks.
    #' @param weight Numeric | Character | list[Numeric | Character], optional. Weights to apply.
    #' @param other Numeric, optional.
    #' @param comment Character, optional. Comment.
    initialize = function(where = NULL, use = NULL,weight = NULL, other = NaN, comment = '') {

        # set fields from constructor arguments
        private$..where <- if (is.null(where)){ list() }else{if (typeof(where)=="list"){ where }else{ list(where) }}
        private$..use <- if (is.null(use)){ list() }else{if (typeof(use)=="list"){ use }else{ list(use) }}
        private$..weight <- if(is.null(weight)){ NULL }else{ if (typeof(weight=="list")){lapply(weight, function(x){as.numeric(x)})}else{as.numeric(weight)}}
        private$..other <- other
        private$..comment <- comment

         # perform consistency checks on fields
        if((!is.null(private$..use) && !is.null(private$..weight)) && (length(private$..use) != length(private$..weight))) {
            stop("Must provide same length of 'use' conditions as 'weight' values.")
        }

        # set default weight to 1 if not set otherwise
        if (is.null(private$..weight)){
            private$..weight <- list(length(private$..use) * 1.0)
        }
    },


    #' @description Check if a mask matches a dataframe by verifying if all 'where' conditions match across all rows.
    #'
    #' @param df DataFrame. Dataframe to check for matches.
    #'
    #' @return Logical. If the mask matches the dataframe.
    #' @export
    matches = function(df) {
        for (w in private$..where) {
            if (!all(apply_cond(df, w))){
                return(FALSE)
            }
        }
        return(TRUE)
    },

    #' @description Apply weights to the dataframe
    #'
    #' @param df Dataframe. Dataframe to apply weights on
    #' @return Dataframe. Dataframe with applied weights
    get_weights = function(df, use, weight) {
        ret <- rep(NA, nrow(df))

        # Apply weights where the use condition matches
        for (i in 1:length(use)) {
            cond <- use[[i]]
            w <- weight[i]
            ret[apply_cond(df, cond)] <- w
        }

        # Convert the result to a data frame with the same index as df
        ret <- data.frame(index = rownames(df), weights = ret)

        # return
        return(ret)
     }


  )
)

#' @description Reads YAML files containing mask specifications from multiple databases and returns a list of Mask objects.
#'
#' @param variable Character. Variable to be read.
#'
#' @return List. List with masks for the variable.
#'
#' @export
read_masks <- function(variable) {
    ret <- list()

    for (database_id in names(databases)) {
        fpath <- file.path(databases[[database_id]], 'masks', paste0(paste(unlist(strsplit(variable,split= '\\|')), collapse = '/'), '.yml'))
        if (file.exists(fpath)) {
            if (dir.exists(fpath)) {
                stop(paste("Expected YAML file, but not a file:", fpath))
            }

        ret <- c(ret, lapply(read_yml_file(fpath), function(mask_specs) {Mask$new(mask_specs)}))
        }
    }

  return(ret)
}

File: posted/noslag/_masking.py
"""Define and apply masks for selecting in TEDFs."""

from typing import Callable

import numpy as np
import pandas as pd

from posted import POSTEDException

MaskCondition = str | dict | Callable[[pd.Series], pd.Series]


def _apply_cond(df: pd.DataFrame, cond: MaskCondition) -> pd.Series:
    """Apply condition to dataframe.

    Takes a pandas DataFrame and a condition, which can be a string,
    dictionary, or callable, and applies the condition to the DataFrame
    using `eval` or `apply` accordingly.

    Parameters
    ----------
    df : pd.DataFrame
        A pandas DataFrame containing the data on which the condition
        will be applied.
    cond : MaskCondition
        The condition to be applied on the dataframe. Can be either a
        string, a dictionary, or a callable function.

    Returns
    -------
        pd.DataFrame
            Dataframe evaluated at the mask condition

    """
    if isinstance(cond, str):
        return df.eval(cond)
    elif isinstance(cond, dict):
        cond = " & ".join([f"{key}=='{val}'" for key, val in cond.items()])
        return df.eval(cond)
    elif isinstance(cond, Callable):
        return df.apply(cond, axis=1)
    else:
        raise POSTEDException(f"Unknown condition type: {type(cond)}")


class Mask:
    """Class for defining masks with conditions and weights.

    Parameters
    ----------
    where: MaskCondition | list[MaskCondition], optional
        Where the mask should be applied
    use:  MaskCondition | list[MaskCondition], optional
        Condition on where to use the masks
    weight: None | float | str | list[float | str], optional
        Weights to apply
    other: float, optional

    comment: str, optional
            Comment

    """

    def __init__(
        self,
        where: None | MaskCondition | list[MaskCondition] = None,
        use: None | MaskCondition | list[MaskCondition] = None,
        weight: None | float | str | list[float | str] = None,
        other: float = np.nan,
        comment: str = "",
    ):
        """Initialise mask object from arguments.

        Set fields from constructor arguments, perform consistency checks on
        fields, set default weight to 1 if not set otherwise.
        """
        self._where: list[MaskCondition] = (
            []
            if where is None
            else where
            if isinstance(where, list)
            else [where]
        )
        self._use: list[MaskCondition] = (
            [] if use is None else use if isinstance(use, list) else [use]
        )
        self._weight: list[float] = (
            None
            if weight is None
            else [float(w) for w in weight]
            if isinstance(weight, list)
            else [float(weight)]
        )
        self._other: float = other
        self._comment: str = comment

        # perform consistency checks on fields
        if self._use and self._weight and len(self._use) != len(self._weight):
            raise Exception(
                "Must provide same length of 'use' conditions as "
                "'weight' values."
            )

        # set default weight to 1 if not set otherwise
        if not self._weight:
            self._weight = len(self._use) * [1.0]

    def matches(self, df: pd.DataFrame) -> bool:
        """Check if a mask matches a dataframe.

        All 'where' conditions match across all rows.

        Parameters
        ----------
        df: pd.Dataframe
            Dataframe to check for matches

        Returns
        -------
            bool
                If the mask matches the dataframe

        """
        for w in self._where:
            if not _apply_cond(df, w).all():
                return False
        return True

    def get_weights(self, df: pd.DataFrame) -> pd.Series:
        """Apply weights to the dataframe.

        Parameters
        ----------
        df: pd.Dataframe
            Dataframe to apply weights on

        Returns
        -------
            pd.DataFrame
                Dataframe with applied weights

        """
        ret = pd.Series(index=df.index, data=self._other)

        # apply weights where the use condition matches
        for u, w in zip(self._use, self._weight):
            ret.loc[_apply_cond(df, u)] = w

        return ret
